Encrypted jpy (#31)
anima-wg/constrained-join-proxy: draft-ietf-anima-constrained-join-proxy.md

CoAP can be run with the Datagram Transport Layer Security (DTLS) {{RFC6347}} as a security protocol for authenticity and confidentiality of the messages.
This is known as the "coaps" scheme.
A constrained version of EST, using Coap and DTLS, is described in {{RFC9148}}.

> The {{I-D.ietf-anima-constrained-voucher}} extends {{RFC9148}} with BRSKI artifacts such as voucher, request voucher, and the protocol extensions for constrained Pledges.

Author: ```suggestion
The {{I-D.ietf-anima-constrained-voucher}} extends {{RFC9148}} with BRSKI artifacts such as voucher, voucher request, and the protocol extensions for constrained Pledges.
```

   1. The context payload.  This is a CBOR byte string. It SHOULD be between 8 and 32 bytes in size.
   2. Content field: containing the DTLS payload as a CBOR byte string.

~~~
    JPY_message =
    [
       pledge_context   : bstr,
       content   : bstr
    ]

~~~
{: #fig-cddl title='CDDL representation of JPY message' align="left"}

The Join Proxy cannot decrypt the DTLS payload and has no knowledge of the transported media type.
The contents are DTLS encrypted.

The context payload is to be reflected by the Registrar when sending reply packets to the Join Proxy.
The context payload is not standardized.
It is to be used by the Join Proxy to record which pledge the traffic came from.

The Join Proxy SHOULD encrypt this context with a symmetric key known only to the Join Proxy.
This key need not persist on a long term basis, and MAY be changed periodically.
The considerations of  {{Section 5.2 of RFC8974}} apply.

This is intended to be identical to the mechanism described in {{Section 7.1 of RFC9031}}.
However, since the CoAP layer is inside of the DTLS layer (which is between the Pledge and the Registrar), it is not possible for the Join Proxy to act as a CoAP proxy.

> A typical context parameter might be constructed with the following structure.

Author: ```suggestion
A typical context parameter might be constructed with the following CDDL structure.
```

> A typical context parameter might be constructed with the following CDDL structure.

Author: ```suggestion
A typical context parameter might be constructed with the following CDDL grammar:
```

(This is illustrative only: the contents are not subject to standardization)

~~~~
    pledge_context_message = [
      family:  uint .bits 1,
      ifindex: uint .bits 8,
      srcport: uint .bits 16,
      iid:     bstr .bits 64,
    ]
~~~~

> This results in a total of 96 bits, or 12 bytes.
> This fits nicely into a single AES128 CBC block for instance, resulting in a 16 byte context message.

Author: ```suggestion
This results in a total of 96 bits, or 12 bytes.
The structure stores the srcport, the originating IPv6 Link-Local address, the IPv4/IPv6 family (as a single bit) and an ifindex to provide the link-local scope.
This fits nicely into a single AES128 CBC block for instance, resulting in a 16 byte context message.
```

The Join Proxy MUST maintain the same context block for all communications from the same pledge.
This implies that any encryption key either does not change during the communication, or that when it does, it is acceptable to break any onboarding connections which have not yet completed.
If using a construct like above, this should not be a problem to do without maintaining state.

> Note: only the lower 64-bits of the origin IP need to be recorded, because they are all IPv6 Link-Local addresses, so the upper 64-bits are just "fe80::"

Author: ```suggestion
Note: when IPv6 is used only the lower 64-bits of the origin IP need to be recorded, because they are all IPv6 Link-Local addresses, so the upper 64-bits are just "fe80::". For IPv4, a Link-Local IPv4 address {{?RFC3927}} would be used, and it would fit into 64-bits.
```

On media where the IID is not 64-bits, a different arrangement will be necessary.

> The Join Proxy SHOULD use the same UDP source port for all traffic from all Pledges.

Author: ```suggestion
For the JPY messages relayed to the Registrar, the Join Proxy SHOULD use the same UDP source port for the JPY messages related to all pledges.
```

A Join Proxy MAY change the UDP source port, but doing so creates more local state.
A Join Proxy with multiple CPUs (unlikely in a constrained system, but possible in the future) could, for instance, use different source port numbers to demultiplex connections across CPUs.

### Processing by Registrar

> On reception by the Registrar, the Registrar MUST verify that the number of array elements is 2 or more.

Author: ```suggestion
On reception of a JPY message by the Registrar, the Registrar MUST verify that the number of array elements is 2 or more.
```

The pledge_content field must be provided as input to a DTLS library {{RFC9147}}, which along with the 5-tuple of the UDP connection provides enough context for the Registrar to pick an appropriate context.
Note that the socket will need to be used for multiple DTLS flows, which is atypical for how DTLS usually uses sockets.
The context can be used to select an appropriate DTLS context, as DTLS headers do not contain any kind of equivalent to an IPsec SPI number.

> (This can usually be done in an OpenSSL by use of a custom "BIO" driver)
> 
> The context needs to be stored within the DTLS context, and when DTLS records need to be sent, then the context needs to be prepended.

Author: ```suggestion
The pledge\_context\_message needs to be linked to the DTLS context, and when DTLS records need to be sent, then the pledge\_context\_message needs to be prepended to the data that is sent.
```

## Stateless Message structure

The JPY message is constructed as a payload directly above UDP.
There is no CoAP or DTLS layer as both are within the relayed payload.

Header and Contents fields together are consist of one CBOR {{RFC8949}} array of 2 elements, explained in CDDL {{RFC8610}}:

   1. The context payload.  This is a CBOR byte string. It SHOULD be between 8 and 32 bytes in size.
   2. Content field: containing the DTLS payload as a CBOR byte string.

~~~
    JPY_message =
    [

>        context   : bstr,

Author: ```suggestion
       pledge_context   : bstr,
```

       content   : bstr
    ]

~~~
{: #fig-cddl title='CDDL representation of JPY message' align="left"}

The Join Proxy cannot decrypt the DTLS payload and has no knowledge of the transported media type.
The contents are DTLS encrypted.

The context payload is to be reflected by the Registrar when sending reply packets to the Join Proxy.
The context payload is not standardized.
It is to be used by the Join Proxy to record which pledge the traffic came from.

The Join Proxy SHOULD encrypt this context with a symmetric key known only to the Join Proxy.
This key need not persist on a long term basis, and MAY be changed periodically.
The considerations of  {{Section 5.2 of RFC8974}} apply.

This is intended to be identical to the mechanism described in {{Section 7.1 of RFC9031}}.
However, since the CoAP layer is inside of the DTLS layer (which is between the Pledge and the Registrar), it is not possible for the Join Proxy to act as a CoAP proxy.

A typical context parameter might be constructed with the following structure.
(This is illustrative only: the contents are not subject to standardization)

~~~~

>     context_message = [

Author: ```suggestion
    pledge_context_message = [
```

The Join Proxy MUST maintain the same context block for all communications from the same pledge.
This implies that any encryption key either does not change during the communication, or that when it does, it is acceptable to break any onboarding connections which have not yet completed.
If using a construct like above, this should not be a problem to do without maintaining state.

Note: only the lower 64-bits of the origin IP need to be recorded, because they are all IPv6 Link-Local addresses, so the upper 64-bits are just "fe80::"
On media where the IID is not 64-bits, a different arrangement will be necessary.

The Join Proxy SHOULD use the same UDP source port for all traffic from all Pledges.
A Join Proxy MAY change the UDP source port, but doing so creates more local state.
A Join Proxy with multiple CPUs (unlikely in a constrained system, but possible in the future) could, for instance, use different source port numbers to demultiplex connections across CPUs.

### Processing by Registrar

On reception by the Registrar, the Registrar MUST verify that the number of array elements is 2 or more.

> The content field must be provided as input to a DTLS library {{RFC9147}}.

Author: ```suggestion
The pledge_content field must be provided as input to a DTLS library {{RFC9147}}, which along with the 5-tuple of the UDP connection provides enough context for the Registrar to pick an appropriate context.
```

> The origin IP address and UDP port number also needs to be recorded, or a connected socket may be created.

Author: ```suggestion
```
